contracts/modules/governance/MemberRoles.sol

```diff
@@ -129,125 +109,125 @@ contract MemberRoles is IMemberRoles, Governed, LegacyMasterAware {
-  function updateRole(//solhint-disable-line
+  /// Assign or Delete a member from a specific role.
+  ///
+  /// @param _memberAddress  Address of the scoped member.
+  /// @param _roleId         The role id to update.
+  /// @param _active         True if we want to assign this role to member or false otherwise.
+  function updateRole(
     address _memberAddress,
     uint _roleId,
     bool _active
-  )
-  public
-  checkRoleAuthority(_roleId) {
+  ) public checkRoleAuthority(_roleId) {
     _updateRole(_memberAddress, _roleId, _active);
   }
 
-  /**
-   * @dev to add members before launch
-   * @param userArray is list of addresses of members
-   * @param tokens is list of tokens minted for each array element
-   */
-  function addMembersBeforeLaunch(address[] memory userArray, uint[] memory tokens) public onlyOwner {
-    require(!launched);
-
-    for (uint i = 0; i < userArray.length; i++) {
-      require(!ms.isMember(userArray[i]));
-      tc.addToWhitelist(userArray[i]);
-      _updateRole(userArray[i], uint(Role.Member), true);
-      tc.mint(userArray[i], tokens[i]);
-    }
-    launched = true;
-    launchedOn = block.timestamp;
+  /// Verifies if a signed message was signed by kycAuthAddress.
+  ///
+  /// @param hash  The hash of the message that is signed.
+  /// @param v     The recovery id.
+  /// @param r     The r argument from the ECDSA signature.
+  /// @param s     The s argument from the ECDSA signature.
+  function isValidSignature(bytes32 hash, uint8 v, bytes32 r, bytes32 s) public view returns (bool) {
+    bytes memory prefix = "\x19Ethereum Signed Message:\n32";
+    bytes32 prefixedHash = keccak256(abi.encodePacked(prefix, hash));
+    address signer = ecrecover(prefixedHash, v, r, s);
+    return (signer == kycAuthAddress);
+  }
+
+  /// Finalises the sign up process by allowing the user to pay the joining fee and marks the
+  /// address as a member.
+  ///
+  /// @dev A signature is required to verify if the address was approved. It is verified against
+  /// the address and a nonce (incremented if a new signature is required for the same address).
+  ///
+  /// @param _userAddress  The address of the user for whom the joining fee is paid.
+  /// @param data          Encoded data containing the nonce and signature.
+  function signUp(
+    address _userAddress,
+    bytes calldata data
+  ) public override payable {
+    require(_userAddress != address(0), "MemberRoles: Address 0 cannot be used");
+    require(!ms.isPause(), "MemberRoles: Emergency pause applied");
+    require(!isMember(_userAddress), "MemberRoles: This address is already a member");
+    require(
+      msg.value == joiningFee,
+      "MemberRoles: The transaction value should equal to the joining fee"
+    );
 
-  }
+    // Extract the nonce and the compact signature representation.
+    // See: https://eips.ethereum.org/EIPS/eip-2098
+    (uint nonce) = abi.decode(data[:32], (uint));
+    bytes memory signature = data[32:96];
```

Comment:
I could do that but if we'll ever want to remove them we'll have the two types in the function parameters forever to avoid breaking changes. The intention here is to avoid bricking any integrations relying on the `signUp` function that would be caused by changes in the function selector.

A similar pattern which gives us this flexibility already exists in Gateway.sol in functions such as `getCoverPrice`, `buyCover` etc. This means that if we reserve this last parameter to be an arbitrary array of byters, we can change the behavior of the function without changing the signature. There would also be no need for an additional gateway contract which essentially does exactly that.